diff --git a/lib/tpm2_eventlog_yaml.c b/lib/tpm2_eventlog_yaml.c
index 5cb7e182..b9ca1e18 100644
--- a/lib/tpm2_eventlog_yaml.c
+++ b/lib/tpm2_eventlog_yaml.c
@@ -5,6 +5,7 @@
 #include <stdlib.h>
 #include <string.h>
 #include <uchar.h>
+#include <wchar.h>
 
 #include <tss2/tss2_tpm2_types.h>
 
@@ -23,6 +24,38 @@
 #include <efivar/efivar.h>
 #endif
 
+static size_t c16rtomb(char *restrict s, char16_t c16, mbstate_t *restrict ps)
+{
+        static unsigned internal_state;
+        if (!ps) ps = (void *)&internal_state;
+        unsigned *x = (unsigned *)ps;
+        wchar_t wc;
+
+        if (!s) {
+                if (*x) goto ilseq;
+                return 1;
+        }
+
+        if (!*x && c16 - 0xd800u < 0x400) {
+                *x = c16 - 0xd7c0 << 10;
+                return 0;
+        }
+
+        if (*x) {
+                if (c16 - 0xdc00u >= 0x400) goto ilseq;
+                else wc = *x + c16 - 0xdc00;
+                *x = 0;
+        } else {
+                wc = c16;
+        }
+        return wcrtomb(s, wc, 0);
+
+ilseq:
+        *x = 0;
+        errno = EILSEQ;
+        return -1;
+}
+
 static void guid_unparse_lower(EFI_GUID guid, char guid_buf[37]) {
 
     snprintf(guid_buf, 37, "%08x-%04x-%04x-%02x%02x-%02x%02x%02x%02x%02x%02x",
